
/*
 * file: config.proto
 * purpose: message format for exchange with configuration daemon
 * author: Tobias Simon, Ilmenau University of Technology
 */


message Value
{
   enum Type
   {
      STR = 0;
      INT = 1;
      DBL = 2;
      BOOL = 3;
   }
   
   required Type type = 1;
   optional string str_val = 2;
   optional int32 int_val = 3;
   optional double dbl_val = 4;
   optional bool bool_val = 5;
}


message CtrlReq
{
   enum Type
   {
      GET = 0; /* get parameter */
      SET = 1; /* set parameter to value */
      PERSIST = 2; /* persist all parameters */
   }

   required Type type = 1;
   optional string id = 2;
   optional Value val = 3;
}


message CtrlRep
{
   enum Status
   {
      OK = 0;
      PARAM_UNKNOWN = 1;
      TYPE_MISMATCH = 2;
      MALFORMED_ID = 3;
      IO_ERROR = 4;
   }

   required Status status = 1;
   optional Value val = 2;
}


/* 
 * NOTE: there is no dedicated protobuf message for param publishing using ZMQ pub/sub;
 *       this data consists of a PLAIN null-terminated string, followed by a "Value" protobuf message
 *
 * pseudo-message PubData
 * {
 *    required string id = 1;
 *    required Value val = 2;
 * }
 */

